### ISSUE TITLE
wgsl: full reference to module-scope variable is uniform, the value stored there may not be

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#3602]
Spawned from the discussion at: https://github.com/gpuweb/gpuweb/issues/2586#issuecomment-1319082038 

Example:
```rust
var<workgroup> w: array<i32,10>;

@compute @workgroup_size(10)
fn foo(@builtin(local_invocation_index) i:u32) {
    // The *reference* w is uniform.
    let pw = &w;    // should be considered uniform
    let v0 = w[0];  // should be considered non-uniform because of mutability.
    let vi = w[i];  // should be considered non-uniform because of nonuniform indexing *and* mutability
    let p0 = &w[0]; // should be considered uniform because it's an address combining uniform things.
    let pi = &w[i]; // should be considered non-uniform
    let vp0 = *p0; // should be considered non-uniform because of mutability.
    let vp1 = *pi; // non-uniform
}
```

In short: 
* full reference to a module-scope variable is uniform
* address calculation combining only uniform things should also be uniform
* applying LoadRule to a mutable module-scope reference yields MayBeNonUniform
* applying LoadRule to immutable thing yields Uniform

However, uniformity analysis currently says
* full reference to module-scope variable in "lvalue position" is MayBeNonUniform.

This matters because with the current formulation we can't ever say that a pointer to a workgroup variable is uniform.
And that's what need for workgroupUniformLoad. See #2586

**Comment by dneto0** [comment]
In the above we want uniformity analysis to say:

* `pw` is uniform  It's a "full reference" to the module-scope variable

That's what we need to make uniformWorkgroupLoad to work.

**Comment by dneto0** [comment]
The explanatory text already says we're trying to distinguish between the reference-memory-view (address) use vs. the after-LoadRule use.

> We have a similar set of rules for expressions in left-value positions, that we denote by LValue: (CF, E) => (CF, L). Instead of computing the node which corresponds to the uniformity of the value, it computes the node which corresponds to the uniformity of the variable we are addressing.

So the intent was there, but the details seemed to have gotten it wrong.

**Comment by dneto0** [comment]
> The explanatory text already says we're trying to distinguish between the reference-memory-view (address) use vs. the after-LoadRule use.
> 
> > We have a similar set of rules for expressions in left-value positions, that we denote by LValue: (CF, E) => (CF, L). Instead of computing the node which corresponds to the uniformity of the value, it computes the node which corresponds to the uniformity of the variable we are addressing.
> 
> So the intent was there, but the details seemed to have gotten it wrong.

Actually, not quite.  The phrasing here"uniformity of the variable we're addressing" really means the "uniformity of the contents of the variable we're addressing", i.e. it's about the uniformity of the *contents*.

So, the analysis doesn't treat uniformity of the address, as it were, of the memory view separately from the uniformity of the value stored in the memory locations.

**Comment by dneto0** [comment]
I've been reading the uniformity sections to see how to address this.

My current idea is that memory view types (references, pointers) actually correspond to two values: the address of the memory locations, and the contents of those memory locations.  An expression computing the address can be uniform, but the contents non-uniform.  That's the case prompting this bug.

So, in a value analysis the value node for a memory view type is a pair `(view, stored_value)`.  When the LoadRule takes effect, only hook up the stored_value component.  Otherwise, pass along both.

For drafting purposes we can have the convention that for non-memory-view types, value analysis produces a pair with a trivial memory-view component:  (&perp;, V)
Alternately, define a function StoredValNode so that:
*  for memory view  analysis:   StoredValNode( view, stored_value ) = stored_value
*  other value analysis:     StoredValNode(V) = V

A technical problem arises, in that we say:

> If an expression is executed in uniform control flow, and all invocations compute the same value, it is said to be a uniform value.

That kind of breaks down for variables in function and private address spaces, because each invocation has its own copy.
We'd need to wordsmith it, but it's doable.

**Comment by dneto0** [comment]
Fixed by #3586

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-11-29-minutes#wgsl-full-reference-to-module-scope-variable-is-uniform-the-value-stored-there-may-not-be-3602">WGSL 2022-11-29 Minutes</a></summary>

* (DN:  I filed this when thinking through options for workgroupUniformLoad.  I don’t think it’s needed otherwise, and would complicate the spec a lot.)
* DN: AB drafted change to go along with workgroup Uniform Layout. Taking JBs suggestion to key off of load rule to do analysis. Agree that is a good thing to do, trying to draft correctly. PR is up and approved but questions from JP. Nothing to do in group right now. Weeks ago said doing workgroupUniformLoad, just need to trust us to write down the correct uniformity rules.
* KG: PR number?
* DN: 3586 [https://github.com/gpuweb/gpuweb/pull/3586](https://github.com/gpuweb/gpuweb/pull/3586) 
* KG: **Sounds good for now**.

</details>
